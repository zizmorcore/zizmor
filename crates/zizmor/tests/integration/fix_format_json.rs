//! Tests for the --fix-format json functionality.
//!
//! These tests verify that the JSON Patch output is correctly formatted
//! and contains the expected operations for different types of fixes.

use crate::common::{input_under_test, zizmor};
use anyhow::Result;

/// Test JSON Patch output for artipacked fixes (persist-credentials issue)
#[test]
fn test_artipacked_fixes_json_patch() -> Result<()> {
    insta::assert_snapshot!(
        zizmor()
            .input(input_under_test("artipacked.yml"))
            .args(["--fix=all", "--fix-format=json", "--quiet"])
            .run()?
    );

    Ok(())
}

/// Test JSON Patch output for template injection fixes
#[test]
fn test_template_injection_fixes_json_patch() -> Result<()> {
    insta::assert_snapshot!(
        zizmor()
            .input(input_under_test("template-injection.yml"))
            .args(["--fix=all", "--fix-format=json", "--quiet"])
            .run()?
    );

    Ok(())
}

/// Test JSON Patch output for cache poisoning fixes
#[test]
fn test_cache_poisoning_fixes_json_patch() -> Result<()> {
    insta::assert_snapshot!(
        zizmor()
            .input(input_under_test("fix-scenarios/cache-poisoning-fixes.yml"))
            .args(["--fix=all", "--fix-format=json", "--quiet"])
            .run()?
    );

    Ok(())
}

/// Test JSON Patch output for unsound condition fixes
#[test]
fn test_unsound_condition_fixes_json_patch() -> Result<()> {
    insta::assert_snapshot!(
        zizmor()
            .input(input_under_test(
                "fix-scenarios/unsound-condition-fixes.yml"
            ))
            .args(["--fix=all", "--fix-format=json", "--quiet"])
            .run()?
    );

    Ok(())
}

/// Test JSON Patch output for mixed fix types
#[test]
fn test_mixed_fixes_json_patch() -> Result<()> {
    insta::assert_snapshot!(
        zizmor()
            .input(input_under_test("fix-scenarios/mixed-fixes.yml"))
            .args(["--fix=all", "--fix-format=json", "--quiet"])
            .run()?
    );

    Ok(())
}

/// Test that no fixes produces empty output (only version banner)
#[test]
fn test_no_fixes_produces_empty_output() -> Result<()> {
    insta::assert_snapshot!(
        zizmor()
            .input(input_under_test("fix-scenarios/no-fixes.yml"))
            .args(["--fix=all", "--fix-format=json", "--quiet"])
            .run()?
    );

    Ok(())
}

/// Test that different fix modes work with JSON format
#[test]
fn test_fix_modes_with_json() -> Result<()> {
    // Test safe fixes only
    insta::assert_snapshot!(
        "safe_fixes_only",
        zizmor()
            .input(input_under_test("fix-scenarios/mixed-fixes.yml"))
            .args(["--fix=safe", "--fix-format=json", "--quiet"])
            .run()?
    );

    // Test unsafe fixes only
    insta::assert_snapshot!(
        "unsafe_fixes_only",
        zizmor()
            .input(input_under_test("fix-scenarios/mixed-fixes.yml"))
            .args(["--fix=unsafe-only", "--fix-format=json", "--quiet"])
            .run()?
    );

    // Test all fixes
    insta::assert_snapshot!(
        "all_fixes",
        zizmor()
            .input(input_under_test("fix-scenarios/mixed-fixes.yml"))
            .args(["--fix=all", "--fix-format=json", "--quiet"])
            .run()?
    );

    Ok(())
}

/// Test that multiple files produce multiple patches in deterministic order
#[test]
fn test_multiple_files_produce_multiple_patches() -> Result<()> {
    // Use files with predictable, simple fixes to ensure deterministic output
    insta::assert_snapshot!(
        zizmor()
            .input(input_under_test("artipacked.yml"))
            .input(input_under_test("template-injection.yml"))
            .args(["--fix=all", "--fix-format=json", "--quiet"])
            .run()?
    );

    Ok(())
}

/// Test that multiple file processing is deterministic
#[test]
fn test_multiple_files_deterministic_output() -> Result<()> {
    // Run the same test twice to ensure deterministic output
    let output1 = zizmor()
        .input(input_under_test("artipacked.yml"))
        .input(input_under_test("template-injection.yml"))
        .args(["--fix=all", "--fix-format=json", "--quiet"])
        .run()?;

    let output2 = zizmor()
        .input(input_under_test("artipacked.yml"))
        .input(input_under_test("template-injection.yml"))
        .args(["--fix=all", "--fix-format=json", "--quiet"])
        .run()?;

    // The outputs should be identical
    assert_eq!(
        output1, output2,
        "Multiple file processing should be deterministic"
    );

    Ok(())
}
