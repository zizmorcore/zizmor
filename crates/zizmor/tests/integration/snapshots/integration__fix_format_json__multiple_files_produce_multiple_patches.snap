---
source: crates/zizmor/tests/integration/fix_format_json.rs
expression: "zizmor().input(input_under_test(\"artipacked.yml\")).input(input_under_test(\"template-injection.yml\")).args([\"--fix=all\",\n\"--fix-format=json\", \"--quiet\"]).run()?"
---
warning[artipacked]: credential persistence through GitHub Actions artifacts
  --> @@INPUT@@:18:9
   |
18 |       - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # tag=v4.2.2
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ does not set persist-credentials: false
   |
   = note: audit confidence → Low
   = note: this finding has an auto-fix

error[template-injection]: code injection via template expansion
  --> @@INPUT@@:19:40
   |
16 |       - uses: actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea # tag=v7.0.1
   |         -------------------------------------------------------------------- action accepts arbitrary code
17 |         with:
18 |           script: |
   |           ------ via this input
19 |             return "doing a thing: ${{ github.event.issue.title }}"
   |                                        ^^^^^^^^^^^^^^^^^^^^^^^^ may expand into attacker-controllable code
   |
   = note: audit confidence → High

4 findings (2 suppressed, 1 fixable): 0 informational, 0 low, 1 medium, 1 high
[
  {
    "file": "@@INPUT@@",
    "operations": [
      {
        "op": "add",
        "path": "/jobs/artipacked/steps/0/with",
        "value": {
          "persist-credentials": false
        }
      }
    ]
  }
]
